### PR TITLE
Strip "words" consisting mainly of punctuation.

### PR DESCRIPTION
**Pull request by btimby**
This adds a post-processing step for binary files. The idea is to remove obvious garbage. Any runs of punctuation, or word chars with two adjacent punctuation chars are stripped. Unit test to verify these rules are followed.

### CODE REVIEW

File: fulltext/__init__.py [hunk 1]
```diff
@@ -71,6 +72,32 @@ def _import_backends():
     LOGGER.info('Loaded backends: %s', ', '.join(BACKENDS.keys()))
 
 
+def is_binary(f):
+    """Return True if binary mode."""
+    # Python 2 makes no distinction.
+    if not PY3:
+        return True
```
Comment (giampaolo) [inline fulltext/__init__.py]: may fail if passed a `codes.open` instance

File: fulltext/__init__.py [hunk 2]
```diff
@@ -71,6 +72,32 @@ def _import_backends():
...
+    if 'b' in getattr(f, 'mode', ''):
+        return True
+
+    # If it has a decoding attribute with a value, it is text mode.
+    if getattr(f, "encoding", None):
+        return False
+
+    # Can we sniff?
+    try:
+        f.seek(0, os.SEEK_CUR)
```
Comment (giampaolo) [inline fulltext/__init__.py]: will this work on Windows? (actually not sure if this is supposed to work on win)
also consider adding a test which makes sure the offset of a file-like obj is correctly reset to its initial position

File: fulltext/__init__.py [hunk 3]
```diff
@@ -98,6 +125,9 @@ def _get_file(backend, f, **kwargs):
     backend's `_get_file()` if one is provided. Otherwise, it will write the
     data to a temporary file and call `_get_path()`.
     """
+    if not is_binary(f):
+        raise AssertionError('File must be opened in binary mode.')
```
Comment (giampaolo) [inline fulltext/__init__.py]: I would use `ValueError` just in case somebody uses `-O` switch

File: fulltext/backends/__csv.py [hunk 4]
```diff
@@ -10,7 +10,13 @@
 
 def readlines(f):
     for line in f.readlines():
-        yield line.decode('utf8')
+        # Decoding may be necessary.
+        try:
+            line = line.decode('utf8')
+        except AttributeError:
+            pass
```
Comment (giampaolo) [inline fulltext/backends/__csv.py]: isn't this unnecessary now that we make sure file is binary up-front?

File: fulltext/__init__.py (hunk 2 above)
Comment (btimby) [reply fulltext/__init__.py]: We do want this to work on Windows. I don't see any notes in Python docs about this not working on Windows.

File: fulltext/__init__.py (hunk 3 above)
Comment (btimby) [reply fulltext/__init__.py]: `-O` disables `assert`, not `AssertionError`.

```
$ python -O
>>> assert 1 == 2, 'Foobar'
>>> raise AssertionError()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

File: fulltext/backends/__csv.py (hunk 4 above)
Comment (btimby) [reply fulltext/backends/__csv.py]: The `try...except` is unecessary.

File: fulltext/__init__.py [hunk 5]
```diff
@@ -71,6 +72,32 @@ def _import_backends():
...
 
+def is_binary(f):
+    """Return True if binary mode."""
+    # If it has a decoding attribute with a value, it is text mode.
+    if getattr(f, "encoding", None):
+        return False
+
+    # Python 2 makes no distinction.
+    if not PY3:
+        return True
```
Comment (giampaolo) [inline fulltext/__init__.py]: on Python 2:
```
>>> io.open('setup.py', 'r', encoding='utf-8').read(1)
u'#'
```

**Comment by giampaolo** [comment]
LGTM

**Pull request merged**